modules/apps/27-interchain-accounts/host/keeper/msg_server.go

```diff
@@ -23,6 +26,40 @@ func NewMsgServerImpl(keeper *Keeper) types.MsgServer {
 	return &msgServer{Keeper: keeper}
 }
 
+// ModuleQuerySafe routes the queries to the keeper's query router if they are module_query_safe.
+// This handler doesn't use the signer.
+func (m msgServer) ModuleQuerySafe(goCtx context.Context, msg *types.MsgModuleQuerySafe) (*types.MsgModuleQuerySafeResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	responses := make([][]byte, len(msg.Requests))
+	for i, query := range msg.Requests {
+		isModuleQuerySafe := slices.Contains(m.mqsWhitelist, query.Path)
+		if !isModuleQuerySafe {
+			return nil, errorsmod.Wrapf(ibcerrors.ErrInvalidRequest, "cannot verify module query safe: %s", query.Path)
+		}
+
+		route := m.queryRouter.Route(query.Path)
+		if route == nil {
+			return nil, errorsmod.Wrapf(ibcerrors.ErrInvalidRequest, "no route to query: %s", query.Path)
+		}
+
+		res, err := route(ctx, &abci.RequestQuery{
+			Path: query.Path,
+			Data: query.Data,
+		})
+		if err != nil {
+			return nil, err
```

Comment:
Will the error returned here include the path of the query? Just thinking that if it's not included then it might be difficult to figure out what query is causing issues if the request includes multiple queries?

Reply:
I'm doing what wasmd and 08-wasm does. I think we should keep it this way:
https://github.com/CosmWasm/wasmd/blob/08567ff20e372e4f4204a91ca64a371538742bed/x/wasm/keeper/query_plugins.go#L363-L365

https://github.com/cosmos/ibc-go/blob/5514d56284135be44dd064a1ffbfe5ae5b607b4a/modules/light-clients/08-wasm/types/querier.go#L154-L156

Reply:
Correct if I'm wrong, but in wasmd and 08-wasm queries are not batched, right? My concern here is if multiple queries are batched and one of them fails here, it would be difficult to figure out what request is the problem if this error doesn't include the path.

Reply:
no harm adding some logs here if we don't want to change the error.

Reply:
I think wasmd and 08-wasm queries are not batched. However, if the contract executes back to back queries, can't the same situation still occur? I'm happy to wrap the error if others agree. Also ok with a log.

Reply:
Logging sounds good. Let's do that for now and if it's not enough we can always come back to wrap the error.

Reply:
added logger